src/lib/Imp.hs

```diff
@@ -269,178 +307,311 @@ toImpHof env (maybeDest, hof) = do
---
--- Based on that observation, we represent destinations using regular Imp atoms, with the
--- restriction that all atoms that terminate blocks appearing in the dest have to represent
--- dests or arrays. This allows us to nicely separate the logical layout, which is
--- encoded in table lambdas, from the physical layout of the data, which is encoded in the
--- array offset calculations that are based on the logical indices (i.e. lambda binders).
---
--- Note that we use a newtype to help distinguish between regular atoms and dests, since
--- it should only be allowed to convert from a dest to a regular atom (by simply issuing a
--- load instruction for each fully offset array), but not in the other direction.
--- Also, Dests should also always be fully simplified (i.e. no beta-redexes are allowed,
--- even under lambdas).
 
-newtype Dest = Dest Atom deriving (Show)
+type Dest = Atom  -- has type `Ref a` for some a
 type WithDest a = (Maybe Dest, a)
 
-deriving via Atom instance HasVars Dest
-
-pattern TabValAbs :: LamExpr -> Atom
-pattern TabValAbs a <- Lam a@(Abs _ (TabArrow, _))
+data DestEnv = DestEnv
+       { allocationInfo :: AllocInfo
+       , enclosingIdxs :: IndexStructure
+       , destDepVars :: Env () }
+
+-- The Cat env carries names for the pointers needed, along with their types and
+-- blocks that compute allocation sizes needed
+type DestM = ReaderT DestEnv (CatT (Env (Type, Block)) Embed)
+
+makeDest :: AllocInfo -> Type -> Embed ([(Binder, Atom)], Dest)
+makeDest allocInfo ty = do
+  (dest, ptrs) <- flip runCatT mempty $ flip runReaderT env $ makeDestRec ty
+  ptrs' <- forM (envPairs ptrs) $ \(v, (ptrTy, numel)) -> do
+    numel' <- emitBlock numel
+    return (Bind (v:>ptrTy), numel')
+  return (ptrs', dest)
+  where env = DestEnv allocInfo mempty mempty
+
+makeDestRec :: Type -> DestM Dest
+makeDestRec ty = case ty of
+  TabTy b bodyTy -> do
+    depVars <- asks destDepVars
+    if not $ null $ freeVars b `envIntersect` depVars
+      then do
+        (dest, ptrs) <- local (\env -> env { enclosingIdxs = mempty
+                                           , destDepVars   = mempty }) $ scoped $
+                          makeDestRec ty
+        makeBoxes (envPairs ptrs) dest
+      else do
+        lam <- buildLam (Bind ("i":> binderAnn b)) TabArrow $ \(Var i) -> do
+          bodyTy' <- substEmbed (b@>Var i) bodyTy
+          withEnclosingIdxs (Bind i) $ makeDestRec bodyTy'
+        return $ Con $ TabRef lam
+  TypeCon def params -> do
+    let dcs = applyDataDefParams def params
+    case dcs of
+      [] -> error "Void type not allowed"
+      [DataConDef _ bs] -> do
+        dests <- makeDataConDest bs
+        return $ DataConRef def params dests
+      _ -> do
+        tag <- rec TagRepTy
+        let dcs' = applyDataDefParams def params
+        contents <- forM dcs' $ \(DataConDef _ bs) -> forM (toList bs) (rec . binderType)
+        return $ Con $ ConRef $ SumAsProd ty tag contents
```

Comment:
Shouldn't this at least check that those are not dependent? I suspect that it will fail with some pretty bad error right now (probably something being out of scope?).